### PR TITLE
PE-66: Load configuration from cm

### PR DESCRIPTION
**Pull request by tmohanta-opennms**
### All Contributors

* [ ] Have you read and followed our [Contribution Guidelines](https://github.com/OpenNMS/opennms/blob/develop/CONTRIBUTING.md)?
* [ ] Have you [made an issue in the OpenNMS issue tracker](https://issues.opennms.org)?<br>If so, you should:
  1. update the title of this PR to be of the format: `${JIRA-ISSUE-NUMBER}: subject of pull request`
  2. update the JIRA link at the bottom of this comment to refer to the real issue number
  3. prefix your commit messages with the issue number, if possible
* [ ] Have you made a comment in that issue which points back to this PR?
* [ ] Have you updated the JIRA link at the bottom of this comment to link to your issue?
* [ ] If this is a new or updated feature, is there documentation for the new behavior?
* [ ] If this is new code, are there unit and/or integration tests?
* [ ] If this PR targets a `foundation-*` branch, does it avoid changing files in `$OPENNMS_HOME/etc/`?

### Pull Request Process

One or more reviewers should be assigned to each PR.

If you know that a particular person is subject matter expert in the area your PR affects, feel free to assign one or more reviewers when you create this PR, otherwise reviewers will be assigned for you.

If you have made additions or changes to the documentation, or if you _need_ documentation for these code changes, please make sure a technical writer has looked it over.

Once the reviewer(s) accept the PR and the branch passes continuous integration, the PR is eligible for merge.

At that time, if you have commit access (are an OpenNMS Group employee or a member of the Order of the Green Polo) you are welcome to merge the PR.
Otherwise, a reviewer can merge it for you.

Thanks for taking time to contribute!

### External References

* JIRA (Issue Tracker): http://issues.opennms.org/browse/PE-66

### CODE REVIEW

File: features/discovery/src/test/java/org/opennms/netmgt/discovery/ConfigurationManagerServiceMock.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+import javax.xml.bind.JAXBException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.Set;
+
+@Component
+public class ConfigurationManagerServiceMock implements ConfigurationManagerService {
```
Comment (freddy61025) [inline features/discovery/src/test/java/org/opennms/netmgt/discovery/ConfigurationManagerServiceMock.java:49]: you should reuse the mock in opennms-dao-mock

File: core/daemon/src/test/java/org/opennms/netmgt/daemon/mock-dao.xml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+	   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	   xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.2.xsd">
+
+	<bean id="discoveryFactory" class="org.opennms.netmgt.config.DiscoveryConfigFactory" />
+	<bean id="configurationManagerService" class="org.opennms.netmgt.discovery.ConfigurationManagerServiceMock" />
```
Comment (freddy61025) [inline core/daemon/src/test/java/org/opennms/netmgt/daemon/mock-dao.xml:7]: please reuse opennms-dao-mock

File: features/discovery/src/test/resources/discovery-configuration-managerservice-mock.xml [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+       xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.2.xsd">
+
+    <bean id="discoveryFactory" class="org.opennms.netmgt.config.DiscoveryConfigFactory" />
+    <bean id="configurationManagerService" class="org.opennms.netmgt.dao.mock.ConfigurationManagerServiceMock" />
```
Comment (freddy61025) [inline features/discovery/src/test/resources/discovery-configuration-managerservice-mock.xml:7]: same as above

File: features/discovery/src/test/resources/mock-dao.xml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<beans xmlns="http://www.springframework.org/schema/beans"
+	   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	   xsi:schemaLocation="http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-4.2.xsd">
+
+	<bean id="discoveryFactory" class="org.opennms.netmgt.config.DiscoveryConfigFactory" />
+	<bean id="configurationManagerService" class="org.opennms.netmgt.discovery.ConfigurationManagerServiceMock" />
```
Comment (freddy61025) [inline features/discovery/src/test/resources/mock-dao.xml:7]: same

File: opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java [hunk 5]
```diff
@@ -113,21 +91,29 @@
...
 
     public DiscoveryConfigFactory (DiscoveryConfiguration config) {
+        super(DiscoveryConfiguration.class, "discovery Configuration");
         m_config = config;
     }
 
+    @PostConstruct
+    public void postConstruct() throws IOException {
+        reload();
+        System.out.println(configurationManagerService);
```
Comment (freddy61025) [inline opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java:114]: remove println

File: opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java [hunk 6]
```diff
@@ -196,15 +193,16 @@ protected void saveXml(final String xml) throws IOException {
...
      * <p>saveConfiguration</p>
      *
      * @param configuration a {@link org.opennms.netmgt.config.discovery.DiscoveryConfiguration} object.
      * @throws java.io.IOException if any.
      */
-    public void saveConfiguration(final DiscoveryConfiguration configuration) throws IOException {
-        getWriteLock().lock();
+  public void saveConfiguration(final DiscoveryConfiguration configuration) throws IOException {
+      this.updateConfig(configuration);
+       /* getWriteLock().lock();
```
Comment (freddy61025) [inline opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java:205]: remove comment out block

File: opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java [hunk 7]
```diff
@@ -144,8 +130,19 @@ public Lock getWriteLock() {
...
+            clearUrlSpecifics();
+            clearExcludeRanges();
+            getInitialSleepTime();
+            getRestartSleepTime();
+            getIntraPacketDelay();
+            getConfiguredAddresses();
+        } catch (final Throwable e) {
+            throw new IOException("An error occurred while validating the configuration: " + e.getMessage(), e);
+        }
+       /* try {
```
Comment (freddy61025) [inline opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java:145]: remove comment out block

File: opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java [hunk 8]
```diff
@@ -183,7 +180,7 @@ public synchronized DiscoveryConfiguration getConfiguration() {
      * @param xml a {@link java.lang.String} object.
      * @throws java.io.IOException if any.
      */
-    protected void saveXml(final String xml) throws IOException {
+/*   protected void saveXml(final String xml) throws IOException {
```
Comment (freddy61025) [inline opennms-config/src/main/java/org/opennms/netmgt/config/DiscoveryConfigFactory.java:183]: remove comment out block

**Review by freddy61025 — COMMENTED** [review]
there is some minor places to fix.

**Comment by freddy61025** [comment]
Also where is your liquibase changelog-cm.xml  ? Is it in other PR ?

**Comment by tmohanta-opennms** [comment]
> Also where is your liquibase changelog-cm.xml ? Is it in other PR ?



> Also where is your liquibase changelog-cm.xml ? Is it in other PR ?

I thought it was not required, But I'll commit it now.

**Review by freddy61025 — APPROVED** [review]

**Pull request merged**